mpcite/doi_builder.py

```diff
@@ -0,0 +1,487 @@
+
+
+        :param adapter: OstiMongoAdapter that keeps track of materials, doi, and other related stores
+        :param osti: osti connection specification
+        :param max_doi_requests: the maximum send size that E-Link allows
+        :param sync: boolean representing whether to sync DOI collection with Elink and Explorer in get_items
+        :param kwargs: other keywords fed into Builder(will be documented as development goes on)
+        """
+        super().__init__(sources=[materials_store, robocrys_store],
+                         targets=[doi_store],
+                         **kwargs)
+        # set connections
+        self.materials_store = materials_store
+        self.robocrys_store = robocrys_store
+        self.doi_store = doi_store
+        self.osti = osti
+        self.elsevier = elsevier
+        self.elink_adapter = ELinkAdapter(osti.elink)
+        self.explorer_adapter = ExplorerAdapter(osti.explorer)
+        self.elsevier_adapter = ElviserAdapter(self.elsevier)
+
+        # set flags
+        self.max_doi_requests = max_doi_requests
+        self.sync = sync
+
+        # set logging
+        self.logger.debug("DOI Builder Succesfully instantiated")
+
+    def get_items(self) -> Iterable:
+        """
+        Get a list of material that requires registration or update with remote servers
+
+        Steps:
+            1. Perform synchronization if sync is True
+                - Note that this is a ONE WAY sync, it only pulls data from remote server and checks whether local data
+                matches
+                - If a mismatch happened
+                    - if a mismatch is caused by there is this material from remote server, but not in our local DOI
+                    collection
+                        - Something VERY wrong happend, log that material and append error field
+                    - If a mismatch is caused by different bibtex or changes in status
+                        - update the DOI record as needed, mark the valid field to False as needed.
+            2. Get all the materials that require an update
+                - A material require an update iff
+                    - It has updated its bibtex description
+                    - its Valid flag is turned to False
+            3. Get all material that require registration, if the flag should_register_new_DOI is set to True
+                - A material needs to be registered iff
+                    - It is not in the local DOI collection and that it is in the materials collection
+
+        Note: This method should NOT send any PUSH request. It will only send GET request to sync, and if remote server
+        needs to be notified of a change in a record, the flag of valid should be turned to False
+
+        Returns:
+            generator of materials to retrieve/build DOI
+        """
+        if self.sync:
+            self.logger.info("Start Syncing with E-Link")
+            try:
+                self.sync_doi_collection()
```

Comment:
So this method would return material IDs to check: `mismatched_ids = self.sync_doi_collection()`

Reply:
I would use the collection to keep track of the state instead of returning it, since i have to set all `valid` fields to false.